aiomisc/pool.py

```diff
@@ -0,0 +1,131 @@
+import asyncio
+from abc import abstractmethod, ABC
+from typing import AsyncContextManager
+
+from .utils import cancel_tasks
+
+
+class ContextManager(AsyncContextManager):
+    __slots__ = "__aenter", "__aexit", "__instance"
+
+    sentinel = object()
+
+    def __init__(self, aenter, aexit):
+        self.__aenter = aenter
+        self.__aexit = aexit
+        self.__instance = self.sentinel
+
+    async def __aenter__(self):
+        if self.__instance is not self.sentinel:
+            raise RuntimeError("Reuse of context manager is not acceptable")
+
+        self.__instance = await self.__aenter()
+        return self.__instance
+
+    async def __aexit__(self, exc_type, exc_value, traceback):
+        await self.__aexit(self.__instance)
+
+
+class PoolBase(ABC):
+    __slots__ = (
+        "_loop",
+        "_recycle_bin",
+        "_instances",
+        "_used",
+        "_recycle",
+        "_tasks",
+    )
+
+    def __init__(self, maxsize=None, recycle=None):
+        assert (
+            recycle is None or recycle > 0
+        ), "recycle should be positive number or None"
+
+        self._loop = asyncio.get_event_loop()
+
+        self._recycle_bin = asyncio.Queue()
+        self._instances = asyncio.Queue(maxsize=maxsize)
+        self._used = set()
+        self._recycle = recycle
+        self._tasks = set()
+
+        self.__create_task(self.__recycler())
+
+    def __create_task(self, coro):
+        task = self._loop.create_task(coro)
+        self._tasks.add(task)
+        task.add_done_callback(self._tasks.remove)
+        return task
+
+    async def __recycler(self):
+        while True:
+            instance = await self._recycle_bin.get()
+            self.__create_task(self._destroy_instance(instance))
+            self._recycle_bin.task_done()
+
+    @abstractmethod
+    async def _create_instance(self):
+        pass
+
+    @abstractmethod
+    async def _destroy_instance(self, instance):
+        pass
+
+    # noinspection PyMethodMayBeStatic,PyUnusedLocal
+    @abstractmethod
+    async def _check_instance(self, instance):
+        return True
+
+    def __len__(self):
+        return len(self._used) + self._instances.qsize()
+
+    async def __acquire(self):
+        if len(self) < self._instances.maxsize:
+            instance = await self._create_instance()
```

Comment:
More than `maxsize` instances can be created here, when `n` parallel tasks `__acquire` at the same time when `len(self)` is less than `self._instances.maxsize` at least by one (`k`).
And it can get worse on the blocking call `await self._instances.put(instance)` below - only `k` of such `__acquire` calls will be unblocked at a time, and `__release` will fail with a `Full` exception.

Reply:
done